Add fields and methods to several primitives

PgBiel opened the pull request:
Closes https://github.com/typst/typst/issues/753.

This PR adds the following fields and methods to primitive types (please feel free to suggest more):

- `length`: fields `em` (get `em` part), `abs` (get `pt` part); methods `cm` (converts `pt` to a floating point `cm`), `mm` (same but to `mm`), `inches` (same but to `inches`).
- `relative length`: fields `ratio` (get ratio part) and `length` (get absolute/`length` part)
- `color`: methods  `kind` (the constructor used: `rgba()` | `cmyk()` | `luma()`), `rgba` (converts to array of rgba values), `hex` (converts any color to a RGBA hex string prefixed by `#`), `cmyk` (converts a luma or cmyk color to an array of cmyk ratio values, or throws an error if it's an rgba color as there is no such conversion), `luma` (returns the value of a luma color, or throws an error if it's an rgba or cmyk color)
- `stroke`: fields `paint`, `thickness`, `cap`, `join`, `dash`, `miter-limit`
- `direction`: now has methods `.axis()` (`"horizontal"` for rtl and ltr, or `"vertical"` for btt and ttb), `.start()` (`left` for ltr, `right` for rtl, `top` for ttb, `bottom` for btt), `.end()` (same idea, but inverted, e.g. `right` for ltr), and `.inv()` (rtl for ltr, ltr for rtl, etc.)
- `alignment`: now has methods `.axis()` (`"horizontal"` for left, right, start, end, center; `"vertical"` for top, bottom, horizon) and `.inv()` (left<->right; top<->bottom; start<->end; center<->center; horizon<->horizon)
- `2d alignment`: now has fields `.x` and `.y` (they return the `x` and `y` components of the 2d alignment, respectively; e.g., `(left + top).x == left` and `(left + top).y == top`), and method `.inv()` (inverts both components; e.g. `(right + bottom).inv() == left + top`, while `(center + horizon).inv() == center + horizon`).
- `angle`: now has the methods `rad` and `deg` (return a float with the conversion of this angle to radians and degrees respectively).

Tests were also added. (Still missing docs for all this stuff, though.)

I'm marking this PR as draft so I can have time to add fields for the new stroke attributes (which were pushed recently - this branch isn't rebased yet); I just wanted to show what I have so far so we can discuss it properly.

Any thoughts? Any other type I should add fields for? Also, I wonder if it would be appropriate if some of these fields were converted to methods (if so, which? Maybe `cm`, `mm`, `inches`, and the color conversion ones?).

PgBiel commented:
By the way, I wonder in which `missing_fields` functions should `#[track_caller]` be added? I added it to all so far, but just wanted to be sure regarding which criteria I should use here

laurmaedje commented:
> By the way, I wonder in which missing_fields functions should #[track_caller] be added?

There are a few stupid track_callers in the code base for some reason. We only need it if the handler panics, not if it returns a string.

PgBiel commented:
> There are a few stupid track_callers in the code base for some reason. We only need it if the handler panics, not if it returns a string.

Ohhh alright, thanks for the insight. I might be updating this PR soon with some changes, and I'll remove the unneeded track_callers from my code lol.

laurmaedje commented on src/geom/length.rs:
```diff
@@ -34,6 +36,20 @@ impl Length {
+
+    /// Get a field from this length.
+    pub fn at(&self, field: &str) -> StrResult<Value> {
+        let round_four_digits = |n| ((n as f64) * 1e4).round() / 1e4;
```
why this?

PgBiel replied:
This was because floating point imprecision was causing issues - 3cm wouldn't be equal to `(3cm).cm` because the latter would return something like `3.000000004`. However, I think it would be better if we turned `.cm` and other conversions into methods where you can specify the precision.

laurmaedje commented on src/eval/value.rs:
```diff
@@ -127,6 +127,16 @@ impl Value {
+            Self::Color(color) => color.at(field),
+            Self::Length(length) => length.at(field),
+            Self::Relative(rel) => rel.at(field),
+            d @ Self::Dyn(dynamic) => {
+                if let Some(stroke) = dynamic.downcast::<PartialStroke>() {
+                    stroke.at(field)
+                } else {
+                    Err(eco_format!("cannot access fields on type {}", d.type_name()))
+                }
+            }
```
```suggestion
            Self::Dyn(dynamic) => {
                if let Some(stroke) = dynamic.downcast::<PartialStroke>() {
                    stroke.at(field)
                } else {
                    Err(eco_format!("cannot access fields on type {}", dynamic.type_name()))
                }
            }
```

laurmaedje commented on src/geom/length.rs:
```diff
@@ -132,3 +148,10 @@ impl Resolve for Length {
 }
+
+/// The missing field access error message.
+#[cold]
+#[track_caller]
```
```suggestion
```

laurmaedje commented on src/geom/paint.rs:
```diff
@@ -1,3 +1,5 @@
+use crate::eval::{Array, Str};
```
crate imports should be below, together with the `super`

laurmaedje commented on src/geom/paint.rs:
```diff
@@ -98,6 +100,24 @@ impl Color {
+            "cmyk" => match self {
+                Self::Cmyk(cmyk) => cmyk.to_array().into(),
+                Self::Luma(luma) => luma.to_cmyk().to_array().into(),
+                _ => Value::None, // no rgba -> cmyk conversion
```
not sure about just returning `none` here to be honest

PgBiel replied:
Yeah, I was worried because you wouldn't be able to prevent an error if the color type were unexpected (note that I'm basically trying to avoid many usages of `repr` here)... Perhaps we can add a `color-type` method?

PgBiel replied:
^As an update to this specifically: I moved this to a method `to-cmyk()`, which now errors on the previously "None" case. In order to be able to prepare for the error, I added the `.kind` field to colors (though I might also move that to a method, but not 100% sure yet).

laurmaedje reviewed (commented):
one thing I'm overall also not sure about is fields vs methods. I feel like most of these should be methods. the thing I can see being fields the most is `color` and `thickness`. maybe the policy could be which are conversion vs. which are just accessors. the `color` field has the additional problem that it won't necessarily be a color anymore once we have gradients.

PgBiel commented:
> one thing I'm overall also not sure about is fields vs methods. I feel like most of these should be methods. the thing I can see being fields the most is `color` and `thickness`.

Yes, for sure! Thanks for the feedback. Indeed, I actually intend to change most of these to methods; I mostly wanted to see how far I could get with those attributes and stuff
I'll try to work on this PR a bit today, but I'll probably only be truly finished in several days (due to upcoming exams haha)

> maybe the policy could be which are conversion vs. which are just accessors.

Makes sense to me.

> the `color` field has the additional problem that it won't necessarily be a color anymore once we have gradients.

Well, I guess the implementation here will depend on how we handle gradients. Perhaps have some sort of `is_solid()` method?

PgBiel commented:
Alright, so, updated the branch to latest `main` (adding fields for the stroke things). I'm still planning to add a bit more stuff (and also moving several fields to methods) before marking this PR as ready for review.

In the last few commits, it's worth saying that I made a few sorta significant changes. In particular, I created `eval/fields.rs`, akin to `eval/methods.rs`, to organize functions related to fields, including a `fields_on` function. I needed to take such a move in order to implement field autocomplete on `ide/`, but there might be a different/better way to do this (feel free to suggest changes here).

To be honest, I'd normally prefer some sort of trait system to be able to have value types implement their fields by themselves, though I acknowledge this would be hard to do (perhaps in a future PR or something, if it ever happens - maybe if the 'methods-as-fields' idea goes forward). So this should work.

PgBiel commented:
Regarding colors, I made the following changes: added the `kind` field to tell different color types apart (rgba vs cmyk vs luma) and the `values` field to return the color's values (i.e., for `rgba`, returns `(R, G, B, A)` as ints; for `cmyk`, returns `(C, M, Y, K)` as ratios; for `luma`, I made it return a unit array `(luma_value,)` for type consistency (I can change this if you want).

Though, I feel like returning the CMYK ratios is hacky, as obtaining the original ratios given is not 100% possible. I think we should just accept both ratios and ints as valid inputs to `cmyk()`, and then always display ints in `repr()` and `.values`, which would fix #787. (Thoughts? This is very trivial to do, but could be made in a separate PR if you prefer.)

Finally, we should probably decide on some documentation format for type fields, preferably similar to their methods. Feel free to send suggestions; I will try to take a look on this later (not too worried about this right now).

laurmaedje commented:
Great to see work on this resuming. If there are many fields, an extra module certainly makes sense although I share your desire for having this more trait based in the future. One big question I have: Didn't we want to make most of these methods?

PgBiel commented:
> One big question I have: Didn't we want to make most of these methods?

Yes! Slowly working on that, haha (sorry, forgot to mention this explicitly in my update). Currently I have only done this for colors so far: `.kind` and `.values` are fields as they are pretty basic and don't require much logic, while all the conversion things - hex, cmyk, rgba - were moved to methods. (`.luma` became unnecessary with `.values`)

Though, I'm guessing that `.kind` could also be moved to a method, as I guess it's sort of generated on the fly. But `.values` IMHO makes sense as a field for me, as it might even make sense to perhaps make it assignable :eyes: .

Either way, I'll be working on moving other types' stuff to methods next ;)

PgBiel commented:
Alright! Today's update:

- Rebased to latest `main` (after all the wonderful bugfixes);
- Changed `precision:` to `digits:` in `.cm()`, `.mm()`, `.inches()` for consistency with `calc.round()`
  - Note that I kept this argument because precision sucks here with typical floats, so `3.345cm == (3.345cm).cm()` wouldn't normally hold, for example (due to some digit at the very end). So I defaulted precision to 10 digits (such that the equality holds) while allowing for 15 (with `digits: 15`), but, ideally, we'd use some sort of BigDecimal library. We can also just return the float straight away and force the user to `calc.round` if that's preferrable.
- New field/method implementations:
  - `direction`: now has methods `.axis()` (`"horizontal"` for rtl and ltr, or `"vertical"` for btt and ttb), `.start()` (`left` for ltr, `right` for rtl, `top` for ttb, `bottom` for btt), `.end()` (same idea, but inverted, e.g. `right` for ltr), and `.inverse()` (rtl for ltr, ltr for rtl, etc.)
    - Note that I omitted the internal `is_positive` method here as I didn't find much use for it for users (/could be confusing), but feel free to tell me if I should expose it here as well.
  - `alignment`: now has methods `.axis()` (`"horizontal"` for left, right, start, end, center; `"vertical"` for top, bottom, horizon) and `.inverse()` (left<->right; top<->bottom; start<->end; center<->center; horizon<->horizon)
    - **Regarding `.axis()`** I made start and end return horizontal as that's how it's done internally, **but** we could wish to error instead as start and end could, I guess, have a vertical axis (if typst ever adds support for this). This kind of future-proofing could be excessive though, which is why I am mentioning this here just for awareness (it's likely fine to keep it as horizontal for now, even if for consistency with how 2d alignment works).
    - **Regarding `.inverse()`:** I considered that, while `start` and `end` can technically be arbitrary, in practice, we shouldn't ever expect them to _not_ be the inverse of each other. (Why would they have the same value, or different axes?)
  - `2d alignment`: now has fields `.horizontal` and `.vertical` (they return the `x` and `y` components of the 2d alignment, respectively; e.g., `(left + top).horizontal == left` and `(left + top).vertical == top`), and method `.inverse()` (inverts both components; e.g. `(right + bottom).inverse() == left + top`, while `(center + horizon).inverse() == center + horizon`).
    - Feel free to tell me if you'd prefer `x` and `y` as the field names (instead of `horizontal` and `vertical`), for example.
 
 And that's mostly all for today (besides other minor changes). (Just sharing this to make it easier to stay in touch with this PR's progress.)

laurmaedje commented:
- I think users can round themselves, the `.cm()` method should just perform the conversion
- `inverse()` sounds a bit like an in-place operation. Perhaps just `inv()`? That would also be consistent with some symbol modifiers.
- You can safely assume the `start` and `end` are horizontal for now, that's a general assumption.
- `start` and `end` are always the inverse of each other
- I prefer `.x` and `.y` over `.horizontal` and `.vertical`

PgBiel commented:
Thanks for the valuable input!! Will work on your suggestions later today. 👍

laurmaedje commented:
@PgBiel do you have plans to continue this?

laurmaedje commented:
I don't think we need to block this on the types rework. The only thing I'm wondering is whether we had ultimate consensus on what things are fields and what are methods?

PgBiel commented:
> I don't think we need to block this on the types rework. The only thing I'm wondering is whether we had ultimate consensus on what things are fields and what are methods?

I think we can define that fields should correspond to fundamental parts of each type which we could reasonably expect to eventually be made modifiable (e.g. it might make sense to do something like `rgb_color.values = (1, 2, 3)`, `length.em = 5`, `stroke.paint = blue`, etc.). For now, I think the current separation is good enough, but I'm open for suggestions.

laurmaedje commented:
I agree for stroke, but disagree for color because it's really an enum and having the `values` with different values and unclear interpretation is weird to me. It's not even really useful if you can't assign individual fields like `.r` for RGB. 

With length I'm a bit less sure because it feels a bit like an implementation detail that `em` is resolved lazily but I can see how it's useful.

Regarding `.cm()`: It does seem a bit unnecessary because you can do the division and it goes a bit against "no two ways to do the same thing" (maybe they aren't really at the same level of abstraction though, it's a bit like `in` and `contains()` or `+` and `Add::add` in Rust). The question also has come up a couple of times. I guess I would be okay either way. Improving the docs sounds good in any case. 

Some field names are not ideal yet, here's my list:
- `length.pt` => `length.abs` 
- `rel.relative` => `rel.ratio` (perhaps we should rename it on the rust side too)
- `rel.absolute` => `rel.length` or `rel.abs`
- `color.values` should be removed in my opinion
- `stroke.line_cap` => `stroke.cap` (all four like the dict, could rename on rust side too)
- `stroke.line_join` => `stroke.join`
- `stroke.dash_pattern` => `stroke.dash`
- `stroke.miter_limit` => `stroke.miter-limit`
- `angle.deg` is missing
- `to-rgba` or just `rgba`? same for `cmyk`

laurmaedje commented:
> Perhaps we could repurpose the to-rgba/cmyk methods to just return an array of values instead

I thought they already did!

> and have some .luma() method which returns the luma value 

seems okay. not sure whether it should interpolate or throw. @reknih opinions?

> Regarding .kind(), it could perhaps be made to return the constructor function instead (e.g. return rgba(), luma(), cmyk()); would comparison work here though (kind() == rgba)?

That makes sense! Comparison _should_ work I think.

> But, at least in principle, I don't see much problem with this.

I'm okay with it.

> I'll likely implement it as .deg().

Sounds good.

PgBiel commented:
@laurmaedje the PR seems to be mostly complete now. What's truly missing is **documentation**, and, for that, there are two blockers (I've added what I could so far though):

1. `stroke`, `direction`, `alignment` and `2d alignment` don't seem to be in `types.md`. Should we add sections for them (to accommodate their fields / methods)?
2. What should be the format to document fields and their respective types for a particular type? I was thinking of something like

```md
# Stroke
(...)

## Fields
### paint
The color that a line with this stroke would use.

- type: color

## Methods
(... none in this case so pretend this isn't here ...)
```

PgBiel commented on crates/typst/src/geom/axes.rs:
```diff
@@ -178,14 +178,19 @@ impl Axis {
     }
+
+    /// A description of this axis' direction.
+    pub fn description(self) -> &'static str {
+        match self {
+            Self::X => "horizontal",
+            Self::Y => "vertical",
+        }
+    }
```
I use this method for all `.axis()` methods on alignment. Thus, they return `"horizontal"` and `"vertical"` (just like the `repr` of axis would, although axis isn't really a Typst value anyway, so that doesn't matter much). Feel free to suggest changing this to `"x"` or `"y"` if you prefer.

PgBiel commented on crates/typst/src/geom/stroke.rs:
```diff
@@ -322,6 +322,7 @@ impl Resolve for DashPattern {
 // https://tex.stackexchange.com/questions/45275/tikz-get-values-for-predefined-dash-patterns
 cast! {
     DashPattern,
+    self => dict! { "array" => self.array, "phase" => self.phase }.into_value(),
```
I made dash pattern (returned via `.dash` field of a `stroke`) be converted to a dict with the same structure as the `DashPattern` struct. I thought of maybe comparing with the values for `"solid"`, `"dotted"` etc. to return those strings if applicable, but I settled with this in the end (it's a simpler solution, at least, and encompasses all cases). Feel free to suggest any changes.

Note that we can't directly test this conversion yet, as you can't mutate fields for now (this would be handled by a follow-up PR), and the `stroke` type only appears directly for the user through expressions like `2pt + blue` - anything more complicated (with a different dash pattern than the default `none`) would be expressed by a dict, whose values the user can retrieve directly.

laurmaedje replied:
You can observe it by doing something like `#rect(stroke: (thickness: 1pt, dash: "solid")).stroke`. Looking at the current output from that, some `Debug` impl is suboptimal (it contains `Some`). I'm not super happy with the stroke setup in any case. A `stroke` constructor function instead of the dictionary would maybe make sense.

laurmaedje replied:
Not returning strings like `"solid"` is the right call. The structure should be uniform and "canoncalized" to to speak. Just like `grid(columns: 2).columns == (auto, auto)`.

PgBiel replied:
> You can observe it by doing something like `#rect(stroke: (thickness: 1pt, dash: "solid")).stroke`.
Oh, didn't think of that. I'll add some tests using this. Thanks!

> Looking at the current output from that, some `Debug` impl is suboptimal (it contains `Some`).

The following lines seem to be the case:
1. Debugging `Option<DashPattern>` directly:
https://github.com/PgBiel/typst/blob/23c558fb5750dc80d01802d6e6fc0081def8642c/crates/typst/src/geom/stroke.rs#L162
2. Deriving `Debug` for `DashLength`:
https://github.com/PgBiel/typst/blob/23c558fb5750dc80d01802d6e6fc0081def8642c/crates/typst/src/geom/stroke.rs#L352

I'll take a look at those. Perhaps Typst should get its own `Debug`-like (e.g. `Repr`) trait at some point though, to avoid this kind of issue (since we can't re-implement `Debug` for `Option`, a built-in type). But that's just a thought, and perhaps a very wrong one :p

> I'm not super happy with the stroke setup in any case. A `stroke` constructor function instead of the dictionary would maybe make sense.

I agree; this could be implemented in the type rework.

laurmaedje replied:
I agree on the `Repr` trait, I've actually implemented it at one point. I don't know what stopped from merging it anymore.

PgBiel replied:
Ok, fixed debugging, and added some more tests too. :+1:

PgBiel commented on crates/typst/src/geom/stroke.rs:
```diff
@@ -382,6 +383,11 @@ impl Resolve for DashLength {
 
 cast! {
     DashLength,
+    self => match self {
+        Self::LineWidth => "dot".into_value(),
+        Self::Length(l) => l.into_value(),
+    },
+
```
Might wanna take a look at this and confirm it's OK. Note that obtaining the actual value on the `LineWidth` case isn't possible in the generic case... But perhaps we can automatically convert the `"dot"` to the stroke's `thickness` when getting its `.dash`? Probably not though

laurmaedje replied:
Not much we can do here I think.

PgBiel commented on crates/typst/src/ide/complete.rs:
```diff
@@ -360,6 +360,15 @@ fn field_access_completions(ctx: &mut CompletionContext, value: &Value) {
         })
     }
 
+    for &field in fields_on(value.type_name()) {
+        ctx.completions.push(Completion {
+            kind: CompletionKind::Constant,
+            label: field.into(),
+            apply: None,
+            detail: None,
+        })
+    }
+
```
(This is fine, right?)

laurmaedje replied:
You could use `value_completion` instead of doing a manual completion. Then, the completion will not just show the name of the field, but also its concrete value.

PgBiel commented on crates/typst/src/eval/library.rs:
```diff
@@ -77,6 +77,12 @@ pub struct LangItems {
     pub enum_item: fn(number: Option<usize>, body: Content) -> Content,
     /// An item in a term list: `/ Term: Details`.
     pub term_item: fn(term: Content, description: Content) -> Content,
+    /// The constructor for the 'rgba' color kind.
+    pub rgb_func: &'static NativeFunc,
+    /// The constructor for the 'cmyk' color kind.
+    pub cmyk_func: &'static NativeFunc,
+    /// The constructor for the 'luma' color kind.
+    pub luma_func: &'static NativeFunc,
```
I chose this over `Func` as it felt like unnecessary wrapping. On my tests, it didn't seem to affect the `LangItems` struct's `Sync` or `Send`-ability at all, so should be fine. Feel free to change this though.

laurmaedje replied:
As long as it compares correcly, it should be fine.

PgBiel replied:
Sounds great. It does seem to compare correctly (I added tests with equality and inequality), so that's nice.

PgBiel reviewed (commented):
Since the code is pretty much already ready for review though, I'll leave some comments on parts I think would be of most interest during a review. Of course, you're free to leave other comments in other parts (and also to change documentation / comments at will).

Feel free to suggest any missing tests as well.

laurmaedje commented on crates/typst/src/eval/fields.rs:
```diff
@@ -0,0 +1,93 @@
+    let not_supported = || Err(no_fields(name));
+    let missing = || Err(missing_field(name, field));
+
+    // Special cases, such as module and dict, are handled by Value itself
```
why not handle everything here?

laurmaedje replied:
Leave it as-is then.

laurmaedje commented on crates/typst/src/eval/methods.rs:
```diff
@@ -24,6 +25,32 @@ pub fn call(
+            "cmyk" => match color {
+                Color::Luma(luma) => luma.to_cmyk().to_array().into_value(),
+                Color::Rgba(_) => {
+                    return Err("cannot obtain CMYK values from color kind 'rgba'")
+                        .at(span)
```
```suggestion
                    bail!(span, "cannot obtain cmyk values from rgba color")
```
same below

laurmaedje commented on crates/typst/src/eval/methods.rs:
```diff
@@ -24,6 +25,32 @@ pub fn call(
+                Color::Rgba(_) => vm.items.rgb_func.into_value(),
+                Color::Cmyk(_) => vm.items.cmyk_func.into_value(),
+            },
+            "hex" => color.to_rgba().to_hex().into_value(),
```
technically, it's hex rgba, but `rgba-hex` is meh and  `rgba().hex()` would only be possible if we had separate types which is overkill in my opinion. so I guess it's fine this way.

PgBiel replied:
yeah, and I guess people around the internet already assume `hex` to be RGBA anyway, so yeah

laurmaedje commented on crates/typst/src/eval/methods.rs:
```diff
@@ -152,6 +179,19 @@ pub fn call(
         },
 
+        Value::Length(length) => match method {
+            "cm" => length.abs.to_cm().into_value(),
```
I think `pt` should be here to and then we drop the `abs` field. There is no reason why `pt` should be special. 

It could be confusing that this just ignores the `em` part, but not much we can do about it (without future lazy magic get rules). But probably `em` should then also be a method here and we drop all fields?

laurmaedje replied:
We could also _keep_ the `abs` field, add `pt()` here and have these functions throw if `em` is not `0`. then you can do `length.abs.cm()`, but `2em.cm()` fails instead of giving  `0.0`.

PgBiel replied:
> We could also _keep_ the `abs` field, add `pt()` here and have these functions throw if `em` is not `0`. then you can do `length.abs.cm()`, but `2em.cm()` fails instead of giving `0.0`.

This sounds more appropriate to me. Perhaps we should add this as a hint as well

PgBiel replied:
Done. Make sure to check the error/hint messages and the docs for any changes you might deem necessary.

laurmaedje commented on crates/typst/src/geom/align.rs:
```diff
@@ -98,6 +98,17 @@ impl GenAlign {
+    /// The inverse alignment.
+    pub const fn inv(self) -> Self {
+        match self {
+            // wouldn't be plausible to have start and end not be inverses of
+            // each other, in terms of text direction.
```
I think this is pretty obvious

PgBiel replied:
fair.

laurmaedje commented on crates/typst/src/geom/color.rs:
```diff
@@ -420,6 +431,14 @@ impl CmykColor {
+
+    /// Converts this color to an array of C, M, Y, K components.
+    pub fn to_array(self) -> Array {
+        // convert to ratio
+        let g = |c| Ratio::new(c as f64 / 255.0);
+
+        array![g(self.c), g(self.m), g(self.y), g(self.k)]
```
```suggestion
        let g = |c| Ratio::new(c as f64 / 255.0);
        array![g(self.c), g(self.m), g(self.y), g(self.k)]
```

laurmaedje commented:
Currently `stroke`, `direction`, `alignment`, and `alignment 2d` link to some function they are often used with. An extra type page would be okay I guess, I didn't do it that way because the `Types` section would grow very large and not all types are equally fundamental. There's also regex for instance. 

Adding a new section to `types.md` seems like a lot of investment (for you in the docs crate and for me in the actual docs generator) into a file I'd like to see removed. Maybe it's best to just document the fields in the opening text for the type as done currently for things that have fields?

With my unmerged type rework, the `Types` section was completely removed, and there were pages for the types in the appropriate categories which makes more sense in my opinion. I didn't finish work on it because other important things came up, but it was also not completely clear how the changes would integrate with other ideas for scripting (e.g. elements as types, lazyness + get rules, etc). Maybe I should still finish it as it would perhaps give a better starting point for future work than we currently have. (And remove the weird distinction between scoped functions and methods.)

PgBiel commented:
> Currently `stroke`, `direction`, `alignment`, and `alignment 2d` link to some function they are often used with. An extra type page would be okay I guess, I didn't do it that way because the `Types` section would grow very large and not all types are equally fundamental. There's also regex for instance.

Hmm. Perhaps I can just add this info in a summarized manner then in the docs for each of those "minor types". E.g. "the method `.inv()` returns the inverse alignment ...". Ideally, we'd have a separate page for each one; however, your proposal of having type pages in their respective categories could probably be more appropriate for this.

> Adding a new section to `types.md` seems like a lot of investment (for you in the docs crate and for me in the actual docs generator) into a file I'd like to see removed. Maybe it's best to just document the fields in the opening text for the type as done currently for things that have fields?

Sure, sounds good to me.

> With my unmerged type rework, the `Types` section was completely removed, and there were pages for the types in the appropriate categories which makes more sense in my opinion. I didn't finish work on it because other important things came up, but it was also not completely clear how the changes would integrate with other ideas for scripting (e.g. elements as types, lazyness + get rules, etc). Maybe I should still finish it as it would perhaps give a better starting point for future work than we currently have. (And remove the weird distinction between scoped functions and methods.)

I agree it probably makes sense. I think there should still be a way to list the currently available scripting types, however, in some sort of "index", as this has been pretty useful to me. (Perhaps this idea could be extended to elements as well, in a separate page.)

PgBiel commented:
Alright, I believe I've added docs for everything now. All that's left is mostly reviewing stuff, adapting docs as needed, etc. I'll be undrafting the PR then, since the main goals were already achieved; feel free to re-draft if you feel that's appropriate, though.

laurmaedje commented:
Thank you! :)
